test/simps.lean

```diff
@@ -214,8 +254,348 @@ run_cmd do
+@[simps {type_md := reducible}] protected def equiv2.symm3 {α β} (f : equiv2 α β) : equiv2 β α :=
+⟨f.inv_fun, f, f.right_inv, f.left_inv⟩
+
+example {α β} (f : equiv2 α β) (y : β) : f.symm y = f.inv_fun y := by simp
+example {α β} (f : equiv2 α β) (x : α) : f.symm.inv_fun x = f x := by simp
+
+example {α β} (f : equiv2 α β) : f.symm.inv_fun = f := by { success_if_fail {simp}, refl }
+example {α β} (f : equiv2 α β) : f.symm3.inv_fun = f := by simp
+
+section
+set_option old_structure_cmd true
+class semigroup (G : Type u) extends has_mul G :=
+(mul_assoc : ∀ a b c : G, a * b * c = a * (b * c))
+end
+
+@[simps] instance {α β} [semigroup α] [semigroup β] : semigroup (α × β) :=
+{ mul := λ x y, (x.1 * y.1, x.2 * y.2),
+  mul_assoc := by { intros, simp only [semigroup.mul_assoc], refl } }
+
+example {α β} [semigroup α] [semigroup β] (x y : α × β) : x * y = (x.1 * y.1, x.2 * y.2) :=
+by simp
+example {α β} [semigroup α] [semigroup β] (x y : α × β) : (x * y).1 = x.1 * y.1 := by simp
+
+structure Semigroup :=
+  (G : Type*)
+  (op : G → G → G)
+  (infix * := op)
+  (op_assoc : ∀ (x y z : G), (x * y) * z = x * (y * z))
+
+namespace Group
+
+instance : has_coe_to_sort Semigroup := ⟨_, Semigroup.G⟩
+instance (G : Semigroup) : has_mul G := ⟨G.op⟩
+
+@[simps] def prod_Semigroup (G H : Semigroup) : Semigroup :=
+{ G := G × H,
+  op := λ x y, (x.1 * y.1, x.2 * y.2),
+  op_assoc := by { intros, dsimp [Group.has_mul], simp [Semigroup.op_assoc] }}
+
+
+end Group
+
+section
+set_option old_structure_cmd true
+class extending_stuff (G : Type u) extends has_mul G, has_zero G, has_neg G, has_subset G :=
+(new_axiom : ∀ x : G, x * - 0 ⊆ - x)
+end
+
+@[simps] def bar : extending_stuff ℕ :=
+{ mul := (*),
+  zero := 0,
+  neg := nat.succ,
+  subset := λ x y, true,
+  new_axiom := λ x, trivial }
+
+section
+local attribute [instance] bar
+example (x : ℕ) : x * - 0 ⊆ - x := by simp
+end
+
+class new_extending_stuff (G : Type u) extends has_mul G, has_zero G, has_neg G, has_subset G :=
+(new_axiom : ∀ x : G, x * - 0 ⊆ - x)
+
+@[simps] def new_bar : new_extending_stuff ℕ :=
+{ mul := (*),
+  zero := 0,
+  neg := nat.succ,
+  subset := λ x y, true,
+  new_axiom := λ x, trivial }
+
+section
+local attribute [instance] new_bar
+example (x : ℕ) : x * - 0 ⊆ - x := by simp
+end
+
+
+end coercing
+
+namespace manual_coercion
+
+structure equiv (α : Sort*) (β : Sort*) :=
+(to_fun    : α → β)
+(inv_fun   : β → α)
+
+local infix ` ≃ `:25 := manual_coercion.equiv
+
+variables {α β γ : Sort*}
+
+instance : has_coe_to_fun $ α ≃ β := ⟨_, equiv.to_fun⟩
+
+def equiv.symm (e : α ≃ β) : β ≃ α := ⟨e.inv_fun, e.to_fun⟩
+
+/-- See Note [custom simps projection] -/
+def equiv.simps.inv_fun (e : α ≃ β) : β → α := e.symm
+@[simps] def equiv.trans (e₁ : α ≃ β) (e₂ : β ≃ γ) : α ≃ γ :=
+⟨e₂ ∘ e₁, e₁.symm ∘ e₂.symm⟩
+
+/-- Composition of equivalences `e₁ : α ≃ β` and `e₂ : β ≃ γ`. -/
+@[simps {simp_rhs := tt}] protected def equiv.trans (e₁ : α ≃ β) (e₂ : β ≃ γ) : α ≃ γ :=
```

Comment:
I'm not sure exactly what's being tested here, so I won't fix myself -- this duplicates the name right above.

Reply:
I think I was just tired last night, and left some declarations there which where just some examples for the doc string (see also previous commit) :)